### PR TITLE
Add initial event wildcard support

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
Add EventWildcard support
-- Add IsEventPathSupersetOf in ClusterInfo which is used to check if the concretePath from logged Events is part of interested Path from read or subscribe interaction.
-- Update ReadClient/Handler to generate/process wildcard event path.
-- Cleanup EventPathParams, apply same pattern to AttributePathParams as well.

#### Change overview
See above

#### Testing
Add unit test in TestClusterInfo to cover all supported EventPath with positive/negative test cases.
Update TestEventLogging to cover wildcard eventId.
Update cirque IM test to cover wildcard eventId.
Update to test with one wildcard eventId which would be interpreted into two paths in handler side in unit roundtrip test.

### CODE REVIEW

File: src/app/EventManagement.cpp [hunk 1]
```diff
@@ -610,20 +610,20 @@ CHIP_ERROR EventManagement::CopyEvent(const TLVReader & aReader, TLVWriter & aWr
...
+    ClusterInfo * interestedEventPath = eventLoadOutContext->mpInterestedEventPaths;
     if (eventLoadOutContext->mCurrentEventNumber < eventLoadOutContext->mStartingEventNumber)
     {
         return false;
     }
-    while (interestedEventPaths != nullptr)
+    ConcreteEventPath path(event.mEndpointId, event.mClusterId, event.mEventId);
+    while (interestedEventPath != nullptr)
     {
         // TODO: Support wildcard event path
```
Comment (erjiaqing) [inline src/app/EventManagement.cpp:621]: This TODO can be removed now?

File: src/app/EventManagement.cpp [hunk 2]
```diff
@@ -610,20 +610,20 @@ CHIP_ERROR EventManagement::CopyEvent(const TLVReader & aReader, TLVWriter & aWr
 
 static bool IsInterestedEventPaths(EventLoadOutContext * eventLoadOutContext, const EventEnvelopeContext & event)
 {
-    ClusterInfo * interestedEventPaths = eventLoadOutContext->mpInterestedEventPaths;
+    ClusterInfo * interestedEventPath = eventLoadOutContext->mpInterestedEventPaths;
```
Comment (erjiaqing) [inline src/app/EventManagement.cpp:613]:
```suggestion
for (auto * interestedPath = eventLoadOutContext->mpInterestedEventPaths; interestedPath != nullptr; interestedPath = interestedPath->Next)
```

And move to the line after `if`

File: src/app/EventPathParams.cpp [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/EventPathParams.h>
+
+namespace chip {
+namespace app {
+
+CHIP_ERROR EventPathParams::BuildEventPath(EventPathIB::Builder & aBuilder) const
```
Comment (erjiaqing) [inline src/app/EventPathParams.cpp:24]: How about move this to a member function of Builder?

i.e. EventPathIB::Builder::BuildEventPath(EventPathParams...)

File: src/app/tests/TestClusterInfo.cpp [hunk 4]
```diff
@@ -23,9 +23,12 @@
  */
 
 #include <app/ClusterInfo.h>
+#include <app/util/mock/Constants.h>
```
Comment (erjiaqing) [inline src/app/tests/TestClusterInfo.cpp:26]: Seems this test don't need the mocked event data? Some random number / constants should be enough for testing IsEventPathSupersetOf(), IsValid...() etc?

File: src/app/tests/TestReadInteraction.cpp [hunk 5]
```diff
@@ -576,7 +576,6 @@ void TestReadInteraction::TestReadClientGenerateOneEventPaths(nlTestSuite * apSu
     NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
 
     chip::app::EventPathParams eventPathParams[2];
```
Comment (erjiaqing) [inline src/app/tests/TestReadInteraction.cpp:578]: Here is the mock should be used -- please add round trip tests for reading wildcard events.

File: src/app/util/mock/Constants.h [hunk 6]
```diff
@@ -45,5 +45,9 @@ constexpr AttributeId MockClusterId(const uint16_t & id)
     return (0xFFF1'0000 | id);
 }
 
+constexpr AttributeId MockEventId(const uint16_t & id)
```
Comment (erjiaqing) [inline src/app/util/mock/Constants.h:48]:
```suggestion
constexpr EventId MockEventId(const uint16_t & id)
```

**Review by erjiaqing — CHANGES_REQUESTED** [review]

File: src/app/util/mock/Constants.h (hunk 6 above)
Comment (yunhanw-google) [reply src/app/util/mock/Constants.h:48]: Done, thanks

File: src/app/tests/TestClusterInfo.cpp (hunk 4 above)
Comment (yunhanw-google) [reply src/app/tests/TestClusterInfo.cpp:26]: Prefer using the mockEventData here, ConcreteEventPath testPath(kMockEndpoint1, MockClusterId(1), MockEventId(2)); yes, without mockEventData, it should also be ok

File: src/app/EventManagement.cpp (hunk 2 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:613]: Done, thanks

File: src/app/EventManagement.cpp (hunk 1 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:621]: removed, thanks

File: src/app/tests/TestReadInteraction.cpp (hunk 5 above)
Comment (yunhanw-google) [reply src/app/tests/TestReadInteraction.cpp:578]: updated to test with one wildcard eventId which would be interpreted into two paths in handler side

**Review by erjiaqing — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

**Comment by andy31415** [comment]
fast track: created and approved by domain owners. Change has unit tests.

**Review by jmartinez-silabs — APPROVED** [review]

**Comment by andy31415** [comment]
ASAN on darwin seems to have complained:

```
==81569==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffeec40b740 at pc 0x0001038cc1f3 bp 0x7ffeec40a310 sp 0x7ffeec40a308
READ of size 4 at 0x7ffeec40b740 thread T0
    #0 0x1038cc1f2 in chip::app::EventPathParams::HasWildcardClusterId() const EventPathParams.h:44
    #1 0x1038c1b2b in chip::app::EventPathParams::IsValidEventPath() const EventPathParams.h:41
    #2 0x1038c1327 in chip::app::ReadClient::GenerateEventPaths(chip::app::EventPaths::Builder&, chip::app::EventPathParams*, unsigned long) ReadClient.cpp:205
    #3 0x1038bfee2 in chip::app::ReadClient::SendReadRequest(chip::app::ReadPrepareParams&) ReadClient.cpp:145
    #4 0x1038815fe in chip::app::InteractionModelEngine::SendReadRequest(chip::app::ReadPrepareParams&, chip::app::ReadClient::Callback*) InteractionModelEngine.cpp:445
    #5 0x10381bb3f in chip::app::TestReadInteraction::TestReadRoundtrip(_nlTestSuite*, void*) TestReadInteraction.cpp:696
    #6 0x103a6c4da in nlTestRunner nlunit-test.c:213
    #7 0x103839fd6 in TestReadInteraction() TestReadInteraction.cpp:1495
    #8 0x1037f50ea in main TestReadInteraction.driver.cpp:32
    #9 0x7fff68cfecc8 in start+0x0 (libdyld.dylib:x86_64+0x1acc8)

Address 0x7ffeec40b740 is located in stack of thread T0 at offset 320 in frame
    #0 0x10381ab2f in chip::app::TestReadInteraction::TestReadRoundtrip(_nlTestSuite*, void*) TestReadInteraction.cpp:661

  This frame has 10 object(s):
    [32, 56) 'err' (line 663)
    [96, 136) 'delegate' (line 671)
    [176, 200) 'ref.tmp' (line 673)
    [240, 264) 'ref.tmp10' (line 674)
    [304, 316) 'eventPathParams' (line 677) <== Memory access at offset 320 overflows this variable
    [336, 368) 'attributePathParams' (line 681)
    [400, 496) 'readPrepareParams' (line 691)
    [528, 568) 'agg.tmp'
    [608, 632) 'ref.tmp52' (line 696)
    [672, 696) 'ref.tmp57' (line 697)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow EventPathParams.h:44 in chip::app::EventPathParams::HasWildcardClusterId() const
Shadow bytes around the buggy address:
  0x1fffdd881690: f2 f2 f2 f2 00 00 00 f2 f2 f2 f2 f2 f8 f8 f8 f3
  0x1fffdd8816a0: f3 f3 f3 f3 00 00 00 00 00 00 00 00 00 00 00 00
  0x1fffdd8816b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1fffdd8816c0: f1 f1 f1 f1 00 00 00 f2 f2 f2 f2 f2 00 00 00 00
  0x1fffdd8816d0: 00 f2 f2 f2 f2 f2 f8 f8 f8 f2 f2 f2 f2 f2 f8 f8
=>0x1fffdd8816e0: f8 f2 f2 f2 f2 f2 00 04[f2]f2 00 00 00 00 f2 f2
  0x1fffdd8816f0: f2 f2 00 00 00 00 00 00 00 00 00 00 00 00 f2 f2
  0x1fffdd881700: f2 f2 00 00 00 00 00 f2 f2 f2 f2 f2 00 00 00 f2
  0x1fffdd881710: f2 f2 f2 f2 f8 f8 f8 f3 f3 f3 f3 f3 00 00 00 00
  0x1fffdd881720: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1fffdd881730: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==81569==ABORTING
'#0:','TestReadInteraction'
```

will restart it just in case, but this is the first time I see such a thing.

**Pull request merged**